File: 9-2.rb
#!/usr/bin/env ruby

require 'set'

input = File.read('9.input').lines.map(&:strip).map(&:split)

visited = Set.new

knots = [[0, 0], [0, 0], [0, 0], [0, 0], [0, 0], [0, 0], [0, 0], [0, 0], [0, 0], [0, 0]]

input.each do |insn|
  dir = insn[0]
  steps = insn[1].to_i

  steps.times do
    case dir
    when 'U'
      knots[0][1] += 1
    when 'D'
      knots[0][1] -= 1
    when 'L'
      knots[0][0] -= 1
    when 'R'
      knots[0][0] += 1
    end

    9.times do |previous|
      this = previous + 1

      if (knots[previous][0] - knots[this][0]).abs >= 2 or
         (knots[previous][1] - knots[this][1]).abs >= 2
        knots[this][0] += 1 if knots[this][0] < knots[previous][0]
        knots[this][0] -= 1 if knots[this][0] > knots[previous][0]
        knots[this][1] += 1 if knots[this][1] < knots[previous][1]
        knots[this][1] -= 1 if knots[this][1] > knots[previous][1]
      end
    end
    visited.add [knots[9][0], knots[9][1]]
  end
end

print visited.size, "\n"

File: 9-2.py
#!/usr/bin/env python

input = [row.strip().split() for row in open('9.input').readlines()]

visited = set()

knots = [(0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0)]

for insn in input:
    dir = insn[0]
    steps = int(insn[1])

    for _ in range(steps):
        if dir == "U":
            knots[0] = (knots[0][0], knots[0][1] + 1)
        elif dir == "D":
            knots[0] = (knots[0][0], knots[0][1] - 1)
        elif dir == "L":
            knots[0] = (knots[0][0] - 1, knots[0][1])
        elif dir == "R":
            knots[0] = (knots[0][0] + 1, knots[0][1])

        for previous in range(9):
            this = previous + 1

            if abs(knots[previous][0] - knots[this][0]) >= 2 or \
               abs(knots[previous][1] - knots[this][1]) >= 2:
                step_x = 0
                if knots[this][0] < knots[previous][0]:
                    step_x = 1
                elif knots[this][0] > knots[previous][0]:
                    step_x = -1

                step_y = 0
                if knots[this][1] < knots[previous][1]:
                    step_y = 1
                elif knots[this][1] > knots[previous][1]:
                    step_y = -1

                knots[this] = (knots[this][0] + step_x, knots[this][1] + step_y)

        visited.add(knots[9])

print(len(visited))
